olive/common/ort_inference.py

```diff
@@ -51,26 +75,299 @@ def get_ort_inference_session(
         for key, value in extra_session_config.items():
             sess_options.add_session_config_entry(key, value)
 
-    if isinstance(execution_provider, str):
-        execution_provider = [execution_provider]
-    else:
-        # execution providers should be list[str]
-        assert isinstance(execution_provider, list)
-        assert all(isinstance(ep, str) for ep in execution_provider)
-
-    for idx, ep in enumerate(execution_provider):
-        if ep == "QNNExecutionProvider":
+    # execution providers and provider options
+    providers, provider_options = check_and_normalize_provider_args(
+        inference_settings.get("execution_provider"),
+        inference_settings.get("provider_options"),
+        ort.get_available_providers(),
+    )
+    for idx, provider in enumerate(providers):
+        if provider in ["CUDAExecutionProvider", "DmlExecutionProvider"] and device_id is not None:
+            provider_options[idx]["device_id"] = str(device_id)
+        elif provider == "QNNExecutionProvider":
             # add backend_path for QNNExecutionProvider
-            execution_provider[idx] = ("QNNExecutionProvider", {"backend_path": "QnnHtp.dll"})
-            break
+            provider_options[idx]["backend_path"] = "QnnHtp.dll"
+    logger.debug(f"Normalized providers: {providers}, provider_options: {provider_options}")
 
     # dml specific settings
-    if len(execution_provider) >= 1 and execution_provider[0] == "DmlExecutionProvider":
+    if len(providers) >= 1 and providers[0] == "DmlExecutionProvider":
         sess_options.enable_mem_pattern = False
 
-    provider_options = inference_settings.get("provider_options")
-
     # create session
-    return ort.InferenceSession(
-        str(model_path), sess_options=sess_options, providers=execution_provider, provider_options=provider_options
+    session = ort.InferenceSession(
+        str(model_path), sess_options=sess_options, providers=providers, provider_options=provider_options
     )
+    check_ort_fallback(session, providers)
+    # set tuning results for tunable operators (currently only for ROCM EP)
+    tuning_op_result = inference_settings.get("tuning_op_result")
+    if tuning_op_result:
+        assert isinstance(tuning_op_result, list)
+        session.set_tuning_results(tuning_op_result)
+    return session
+
+
+def check_and_normalize_provider_args(
+    providers: Sequence[Union[str, Tuple[str, Dict[Any, Any]]]],
+    provider_options: Sequence[Dict[Any, Any]],
+    available_provider_names: Sequence[str],
+):
+    """Validate the 'providers' and 'provider_options' arguments and returns a normalized version.
+
+    :param providers: Optional sequence of providers in order of decreasing
```

Comment:
It is not related with this PR.
But how to guarantee the order of desc?

Reply:
I think that is decided by the user if/when they provide their own inference settings. For olive workflows, we loop through the candidate EPs so this will only have one member. 

I think the only current use case with multiple providers in this list is for the rocm perf tuning with migraph and rocm eps, @guotuofeng?